File: benchmarks/python/dict_ops.py
"""Dict Operations Benchmark for Python"""

def bench_build(n, iters):
    d = {}
    for _ in range(iters):
        d = {}
        for i in range(n):
            d[i] = i * 7
    return d

def bench_lookup_hit(d, n, iters):
    checksum = 0
    for _ in range(iters):
        for i in range(n):
            v = d.get(i)
            if v is not None:
                checksum += v
    return checksum

def bench_lookup_miss(d, n, iters):
    misses = 0
    for _ in range(iters):
        for i in range(n, n + n):
            if d.get(i) is None:
                misses += 1
    return misses

def bench_remove(n, iters):
    removed = 0
    for _ in range(iters):
        d = {}
        for i in range(n):
            d[i] = i * 7
        for i in range(n):
            del d[i]
            removed += 1
    return removed

def bench_iterate(d, iters):
    total = 0
    for _ in range(iters):
        for k, v in d.items():
            total += v
    return total

if __name__ == "__main__":
    N = 10000
    d = bench_build(N, 100)
    print(f"build: done")
    hit_sum = bench_lookup_hit(d, N, 100)
    print(f"lookup_hit checksum: {hit_sum}")
    miss_count = bench_lookup_miss(d, N, 100)
    print(f"lookup_miss count: {miss_count}")
    rem_count = bench_remove(N, 100)
    print(f"remove count: {rem_count}")
    iter_sum = bench_iterate(d, 1000)
    print(f"iterate sum: {iter_sum}")

File: benchmarks/go/dict_ops.go
package main

import "fmt"

func benchBuild(n int, iters int) map[int]int {
	var d map[int]int
	for iter := 0; iter < iters; iter++ {
		d = make(map[int]int)
		for i := 0; i < n; i++ {
			d[i] = i * 7
		}
	}
	return d
}

func benchLookupHit(d map[int]int, n int, iters int) int {
	checksum := 0
	for iter := 0; iter < iters; iter++ {
		for i := 0; i < n; i++ {
			if v, ok := d[i]; ok {
				checksum += v
			}
		}
	}
	return checksum
}

func benchLookupMiss(d map[int]int, n int, iters int) int {
	misses := 0
	for iter := 0; iter < iters; iter++ {
		for i := n; i < n+n; i++ {
			if _, ok := d[i]; !ok {
				misses++
			}
		}
	}
	return misses
}

func benchRemove(n int, iters int) int {
	removed := 0
	for iter := 0; iter < iters; iter++ {
		d := make(map[int]int)
		for i := 0; i < n; i++ {
			d[i] = i * 7
		}
		for i := 0; i < n; i++ {
			delete(d, i)
			removed++
		}
	}
	return removed
}

func benchIterate(d map[int]int, iters int) int {
	total := 0
	for iter := 0; iter < iters; iter++ {
		for _, v := range d {
			total += v
		}
	}
	return total
}

func main() {
	const N = 10000

	d := benchBuild(N, 100)
	fmt.Println("build: done")

	hitSum := benchLookupHit(d, N, 100)
	fmt.Printf("lookup_hit checksum: %d\n", hitSum)

	missCount := benchLookupMiss(d, N, 100)
	fmt.Printf("lookup_miss count: %d\n", missCount)

	remCount := benchRemove(N, 100)
	fmt.Printf("remove count: %d\n", remCount)

	iterSum := benchIterate(d, 1000)
	fmt.Printf("iterate sum: %d\n", iterSum)
}
